### PR TITLE
chore: update Log4j to version 2.x

### PR DESCRIPTION
**Pull request by Polpetta**
[CO-715]

### CODE REVIEW

File: store/src/main/java/com/zimbra/cs/service/admin/ResetAllLoggers.java [hunk 1]
```diff
@@ -20,12 +23,25 @@
...
+      ZimbraLog.soap.info("Resetting all loggers");
+      for (Log log : LogManager.getAllLoggers()) {
+        log.removeAccountLoggers();
+      }
+    } finally {
+      lock.unlock();
+    }
+    ZimbraSoapContext zsc = getZimbraSoapContext(context);
+    Element response = zsc.createElement(AdminConstants.RESET_ALL_LOGGERS_RESPONSE);
+    return response;
```
Comment (keshavbhatt) [inline store/src/main/java/com/zimbra/cs/service/admin/ResetAllLoggers.java:45]:
```suggestion
    return  zsc.createElement(AdminConstants.RESET_ALL_LOGGERS_RESPONSE);
```

File: store/pom.xml [hunk 2]
```diff
@@ -475,6 +475,14 @@
       <groupId>io.prometheus</groupId>
       <artifactId>simpleclient_hotspot</artifactId>
     </dependency>
+
+    <!-- Old deprecated dependencies -->
+    <dependency>
+      <groupId>log4j</groupId>
+      <artifactId>log4j</artifactId>
+      <version>1.2.17</version>
+    </dependency>
```
Comment (keshavbhatt) [inline store/pom.xml:484]: are we providing both version of log4j in zcs-lib? 
we need to figureout why we still need it.

**Review by keshavbhatt — CHANGES_REQUESTED** [review]
I have added few suggestions. 

I suggest to track all details about what changes this change has caused to existing logging output format and analyze if we really need them like they were in the past. And create action to be taken in future to acomplish them as logging is very crucial and changing the log format will not make everyone happy.

File: store/pom.xml (hunk 2 above)
Comment (Polpetta) [reply store/pom.xml:484]: As a temporary solution yes, we will submit both. The thing is I think it detects some bogus or deps in deps usage

File: soap/src/main/java/com/zimbra/soap/util/Jaxb2Xsds.java [hunk 3]
```diff
@@ -5,102 +5,93 @@
...
 import java.io.File;
 import java.io.IOException;
 import java.util.List;
-
 import javax.xml.bind.JAXBContext;
 import javax.xml.bind.JAXBException;
 import javax.xml.bind.SchemaOutputResolver;
 import javax.xml.transform.Result;
 import javax.xml.transform.stream.StreamResult;
+import org.apache.logging.log4j.Level;
```
[the remark this answers is not in the document]
Comment (Polpetta) [reply soap/src/main/java/com/zimbra/soap/util/Jaxb2Xsds.java:20]: Formatting e optimizing imports puts it back so it seems necessary

File: common/src/main/java/com/zimbra/common/util/ZimbraLog.java [hunk 4]
```diff
@@ -17,948 +17,752 @@
...
+   * Setup log4j for our command line tools.
+   *
+   * <p>If System.getProperty(zimbra.log4j.level) is set then log at that level. Else log at the
+   * specified defaultLevel.
+   *
+   * @deprecated Because with the switch to Log4j2, it is not more possible to use {@code
+   *     PropertyConfigurator}
+   */
+  @Deprecated(since = "23.7")
+  public static void toolSetupLog4j(String defaultLevel, String logFile, boolean showThreads) {
```
[the remark this answers is not in the document]
Comment (Polpetta) [reply common/src/main/java/com/zimbra/common/util/ZimbraLog.java:629]: I don't want to get rid of logic I don't know. I prefer to mark it as deprecated, put this new codebase in production and see what happens. It we see they're not necessary anymore, we can remove them imo

File: common/src/main/java/com/zimbra/common/util/ZimbraPatternLayout.java [hunk 5]
```diff
@@ -5,44 +5,37 @@
...
- * Subclasses Log4J's <tt>PatternLayout</tt> class to add additional support for
- * the <tt>%z</tt> option, which prints the value returned by {@link ZimbraLog#getContextString()}.
- *   
+ * Subclasses Log4J's <tt>PatternLayout</tt> class to add additional support for the <tt>%z</tt>
+ * option, which prints the value returned by {@link ZimbraLog#getContextString()}.
+ *
  * @author bburtin
  */
-public class ZimbraPatternLayout extends PatternLayout {
+public class ZimbraPatternLayout {
```
[the remark this answers is not in the document]
Comment (Polpetta) [reply common/src/main/java/com/zimbra/common/util/ZimbraPatternLayout.java:19]: It is used in `log4j.properties` file configuration

File: common/src/main/java/com/zimbra/common/util/LogManager.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+/**
+ * Utility class for managing {@link Log}
+ *
+ * @author Davide
+ * @since 23.7
+ * @see LogFactory in order to create Logs
+ */
+public class LogManager {
+
```
[the remark this answers is not in the document]
Comment (Polpetta) [reply common/src/main/java/com/zimbra/common/util/LogManager.java:21]: done

File: common/src/main/java/com/zimbra/common/util/LogManager.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  /**
+   * @return the global {@link Map} with the associated name/{@link Log}
+   */
+  static Map<String, Log> getGlobalLogMapper() {
+    return globalLogMapper;
+  }
+
+  public static void setThisLogAndRootToLevel(Logger logger, Level level) {
```
[the remark this answers is not in the document]
Comment (Polpetta) [reply common/src/main/java/com/zimbra/common/util/LogManager.java:58]: done even if I think that javadoc for self-explanatory setters is a overhead

File: common/src/main/java/com/zimbra/common/util/LogFactory.java [hunk 8]
```diff
@@ -5,92 +5,27 @@
...
-
-    private static final ConcurrentMap<String, Log> NAME2LOG = new ConcurrentHashMap<String, Log>();
-
-    private LogFactory() {
-    }
-
-    public synchronized static void init() {
-        PropertyConfigurator.configure(LC.zimbra_log4j_properties.value());
-    }
+public class LogFactory {
```
[the remark this answers is not in the document]
Comment (Polpetta) [reply common/src/main/java/com/zimbra/common/util/LogFactory.java:18]: done

**Comment by Polpetta** [comment]
> I suggest to track all details about what changes this change has caused to existing logging output format and analyze if we really need them like they were in the past.

We have to discover this. The idea is to see together if this patch works or not. Didn't have the chance to try on road let's say. In any case I think that the output will probably change, but I didn't change any `log4j.properties` file that were the ones defining the output

**Comment by Polpetta** [comment]
Ok, I've checked why Log4j 1.2.x is still necessary unfortunately:
```
+-zextras:zm-store:war:23.7.0-SNAPSHOT
  +-org.apache.curator:curator-framework:jar:2.0.1-incubating:compile
    +-org.apache.zookeeper:zookeeper:jar:3.4.5:compile
      +-log4j:log4j:jar:1.2.17:compile

+-zextras:zm-store:war:23.7.0-SNAPSHOT
  +-zextras:store-extra-runtime-dependencies:pom:23.7.0-SNAPSHOT:runtime
    +-log4j:log4j:jar:1.2.17:runtime

+-zextras:zm-store:war:23.7.0-SNAPSHOT
  +-log4j:log4j:jar:1.2.16:compile
```

This is what `goal org.apache.maven.plugins:maven-enforcer-plugin:3.1.0:enforce` on `zm-store` says.

**Comment by frisonisland** [comment]
> Ok, I've checked why Log4j 1.2.x is still necessary unfortunately:
> 
> ```
> +-zextras:zm-store:war:23.7.0-SNAPSHOT
>   +-org.apache.curator:curator-framework:jar:2.0.1-incubating:compile
>     +-org.apache.zookeeper:zookeeper:jar:3.4.5:compile
>       +-log4j:log4j:jar:1.2.17:compile
> 
> +-zextras:zm-store:war:23.7.0-SNAPSHOT
>   +-zextras:store-extra-runtime-dependencies:pom:23.7.0-SNAPSHOT:runtime
>     +-log4j:log4j:jar:1.2.17:runtime
> 
> +-zextras:zm-store:war:23.7.0-SNAPSHOT
>   +-log4j:log4j:jar:1.2.16:compile
> ```
> 
> This is what `goal org.apache.maven.plugins:maven-enforcer-plugin:3.1.0:enforce` on `zm-store` says.

So:
- store-extra-runtime-dependencies are libraries needed at runtime and should be the same you get from zcs-lib in production. We could put log4j 2 in place of 1.2.
- We can remove log4j 1.2 from zm-store and put log4j 2 probably, unless some classes of 1.2 are used somewhere in the code
- Zookeper dependency scare me the most, because we could have runtime issues

I noticed we are declaring log4j dependency as type 'pom' and scope import. Thus zm-store should declare log4j-core and log4-api in place of log4j in its dependencies, same for store-extra-runtime. I will do some changes locally and propose them if I find it fixes the issue.

File: store/pom.xml (hunk 2 above)
Comment (frisonisland) [reply store/pom.xml]: Removed

**Comment by frisonisland** [comment]
Related PRs:
- https://github.com/zextras/carbonio-jython/pull/2
- https://github.com/zextras/carbonio-ldap-utilities/pull/52
- https://github.com/zextras/carbonio-core-utils/pull/52
- https://github.com/zextras/carbonio-zcs-lib/pull/22

**Comment by Polpetta** [comment]
@frisonisland @keshavbhatt related to https://github.com/zextras/carbonio-zcs-lib/pull/22 should we merge this?

**Comment by frisonisland** [comment]
> @frisonisland @keshavbhatt related to [zextras/carbonio-zcs-lib#22](https://github.com/zextras/carbonio-zcs-lib/pull/22) should we merge this?

Not yet, there are some small improvements I would like to add from https://github.com/zextras/carbonio-mailbox/pull/261

**Review by frisonisland — APPROVED** [review]

**Review by keshavbhatt — APPROVED** [review]

**Pull request merged**